pkg/gardenlet/controller/managedseed/managedseed_actuator.go

```diff
@@ -0,0 +1,636 @@
+		},
+	}
+	if err := a.gardenClient.Client().Delete(ctx, seed); client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	// Check if there are any remaining objects associated with the seed
+	if err := a.checkSeedAssociations(ctx, managedSeed.Name); err != nil {
+		return err
+	}
+
+	// Check if the seed still exists
+	if err := a.gardenClient.Client().Get(ctx, kutil.Key(managedSeed.Name), &gardencorev1beta1.Seed{}); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	return fmt.Errorf("seed %q still exists", managedSeed.Name)
+}
+
+func (a *actuator) deployGardenlet(ctx context.Context, shootClient kubernetes.Interface, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot) error {
+	// Decode gardenlet configuration
+	gardenletConfig, err := helper.DecodeGardenletConfigurationExternal(managedSeed.Spec.Gardenlet.Config, false)
+	if err != nil {
+		return err
+	}
+
+	// Check seed spec and create or update the seed secrets if needed
+	if err := a.prepareSeedSpec(ctx, &gardenletConfig.SeedConfig.SeedTemplate.Spec, managedSeed, shoot); err != nil {
+		return err
+	}
+
+	// Prepare gardenlet chart values
+	values, err := a.prepareGardenletChartValues(ctx, shootClient, managedSeed.Spec.Gardenlet.Deployment, gardenletConfig, managedSeed.Name,
+		v1alpha1helper.GetBootstrap(managedSeed.Spec.Gardenlet.Bootstrap), utils.IsTrue(managedSeed.Spec.Gardenlet.MergeWithParent), shoot)
+	if err != nil {
+		return err
+	}
+
+	// Apply gardenlet chart
+	return shootClient.ChartApplier().Apply(ctx, filepath.Join(common.ChartPath, "gardener", "gardenlet"), v1beta1constants.GardenNamespace, "gardenlet", kubernetes.Values(values))
+}
+
+func (a *actuator) deleteGardenlet(ctx context.Context, shootClient kubernetes.Interface, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot) error {
+	// Decode gardenlet configuration
+	gardenletConfig, err := helper.DecodeGardenletConfigurationExternal(managedSeed.Spec.Gardenlet.Config, false)
+	if err != nil {
+		return err
+	}
+
+	// Delete the seed secrets if needed
+	if err := a.unprepareSeedSpec(ctx, &gardenletConfig.SeedConfig.SeedTemplate.Spec, managedSeed); err != nil {
+		return err
+	}
+
+	// Prepare gardenlet chart values
+	values, err := a.prepareGardenletChartValues(ctx, shootClient, managedSeed.Spec.Gardenlet.Deployment, gardenletConfig, managedSeed.Name,
+		v1alpha1helper.GetBootstrap(managedSeed.Spec.Gardenlet.Bootstrap), utils.IsTrue(managedSeed.Spec.Gardenlet.MergeWithParent), shoot)
+	if err != nil {
+		return err
+	}
+
+	// Delete gardenlet chart
+	return shootClient.ChartApplier().Delete(ctx, filepath.Join(common.ChartPath, "gardener", "gardenlet"), v1beta1constants.GardenNamespace, "gardenlet", kubernetes.Values(values))
+}
+
+func (a *actuator) prepareSeedSpec(ctx context.Context, spec *gardencorev1beta1.SeedSpec, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot) error {
```

Comment:
Yes, you are right. I would split it into 2 methods: `checkShoot` for the first half (it checks certain shoot coditions that could not be checked by the admission controller since they require access to the seed) and `createOrUpdateSecrets` for the second half. The "unprepare" method should then become `deleteSecrets`. OK?

Reply:
sgtm 👍

Reply:
Done